### ISSUE TITLE
at least three suggestions for central problems of project per person

### ISSUE THREAD

**Issue by mfount** [mfount/chicken#11]
by 7:30 tonight

**Comment by mfount** [comment]
here's an idea:

Our problem could be to prove the correctness of Timsort, which is the algorithm that Python uses to sort things.

This has the advantages of being easily modular because it requires a few constituent algorithms (see the wikipedia page for Timsort):
- Minrun finding
- Insertion sort
- run length optimization
- merging runs (requiring a data structure for memory and for temporary memory, separately -- a good opportunity for a "functor") using binary sort and something else
- a "galloping mode" that may or may not be too hard

The core functionality could be to implement a verified Timsort on lists of Coq nats. Then, additional functionality could be to make that algorithm work on higher types, as Ben described in #9. Additional functionality could also be to implement a verified version of the *actual* Timsort that's implemented in Python (slightly different than an abstract version that we'd start with).

We could break up the work by having people be "experts" on various sub-algorithms or data structures/interfaces.

What do you think? If Timsort is too complicated, we could pick another adaptive sorting algorithm to verify.

**Comment by mfount** [comment]
The timsort documentation is beautiful: http://bugs.python.org/file4451/timsort.txt

**Comment by bwetherfield** [comment]
I like this idea for some adaptive sorting algorithm. Will look into timsort in particular.

**Comment by bwetherfield** [comment]
re what Michael has just posted to the google group, I can give this a yes for the draft. Even if the sort is changed later, for now this is a good specific structure to put in place, and you can see how it would be modularized.

**Comment by mfount** [comment]
Great. If others concur, I will get started on a draft, posting some TeX source ASAP so that we can get some feedback and other commits on it

**Comment by bwetherfield** [comment]
I also cp'ed outline.md to draftspec.md  in the draft folder, but I have added next to nothing to it...

**Comment by mfount** [comment]
ah, that works, too. OR, maybe should we "cp" outline.md to a google doc so we can collaborate more easily? Then, we can store the results in the repo in the form of a latex doc (which I can convert the google doc into). (Then, people can proofread / propose final edits from the latex doc.)

**Comment by bwetherfield** [comment]
Better.

**Comment by mfount** [comment]
doing that now.

**Comment by bwetherfield** [comment]
sharelatex?

**Comment by amakelov** [comment]
so it seems someone has already tried verifying timsort, and they even
found a bug
http://envisage-project.eu/proving-android-java-and-python-sorting-algorithm-is-broken-and-how-to-fix-it/

**Comment by amakelov** [comment]
(in the implementations, not the algorithm it seems)

**Comment by mfount** [comment]
re sharelatex: I'll give that a shot in a minute; had never heard of it.

alex: yeah, I saw that, which is why I was differentiating between implementing a verification of the algorithm as core functionality and a verification of the Python implementation of the algorithm as extra features (see outline.md for the difference precisely)

**Comment by ChannyClaus** [comment]
Sorry I just got back from dinner. I’ll start reading the thread from the bottom now

**Comment by ChannyClaus** [comment]
Shouldn’t we try to do verifiers for the simpler sorts first? Maybe if the verifiers we write for bubble sorts or insertion sorts work, we can try fancier sorts later. 

Although I feel like I’m being a buzzkill / underachiever, I’d definitely want to start with something easy especially none of us have written anything in coq yet (or has someone?)

**Comment by mfount** [comment]
But the project needs to have an angle, and applying the Chan Principle, the angle of verifying a hybrid sort by verifying its constiuent sorts/ other algorithms is a good angle for the draft that we need tohave done by tomorrow. Yes?

**Comment by amakelov** [comment]
This would make even more sense if parts of the code that proves
correctness of simpler sorts are reusable for later. For example, insertion
sort is a subroutine of timsort right? So it feels like there should be
some significant overlap there

**Comment by mfount** [comment]
yes, I agree. Let's all go to the chat here https://docs.google.com/document/d/1KcS8DIGFv0VLTOzow4PtS6Z293CwY9iQPY27f6WV21s/edit#

**Comment by bwetherfield** [comment]
Go to the google doc Michael made, not the one I made.
